File: mobile/lib/models/call_transcript.dart
/// Call Transcript model matching backend schema
/// 
/// Purpose: Store complete word-by-word record of all calls for history and debugging
/// 
/// Workflow:
/// 1. Every time a participant speaks, create ONE record per listener group
/// 2. Store original_text (what was actually said)
/// 3. Store translated_text (if translation occurred)
/// 4. Use timestamp_ms to reconstruct timeline
class CallTranscript {
  final String id;
  final String callId;
  final String? speakerUserId;
  final String originalLanguage;
  final String originalText;
  final String? translatedText;
  final String? targetLanguage;
  final int? timestampMs;
  final String? audioFilePath;
  final String? originalAudioPath;
  final String? translatedAudioPath;
  final int? sttConfidence;
  final int? translationQuality;
  final String? ttsMethod;
  final int? processingTimeMs;
  final DateTime createdAt;

  // Joined speaker information
  final String? speakerName;
  final String? speakerAvatarUrl;

  CallTranscript({
    required this.id,
    required this.callId,
    this.speakerUserId,
    required this.originalLanguage,
    required this.originalText,
    this.translatedText,
    this.targetLanguage,
    this.timestampMs,
    this.audioFilePath,
    this.originalAudioPath,
    this.translatedAudioPath,
    this.sttConfidence,
    this.translationQuality,
    this.ttsMethod,
    this.processingTimeMs,
    required this.createdAt,
    this.speakerName,
    this.speakerAvatarUrl,
  });

  /// Create CallTranscript from JSON
  factory CallTranscript.fromJson(Map<String, dynamic> json) {
    return CallTranscript(
      id: json['id'],
      callId: json['call_id'],
      speakerUserId: json['speaker_user_id'] ?? json['speaker_id'],
      originalLanguage: json['original_language'] ?? json['language'] ?? 'he',
      originalText: json['original_text'] ?? json['text'] ?? '',
      translatedText: json['translated_text'] ?? json['translated'],
      targetLanguage: json['target_language'],
      timestampMs: json['timestamp_ms'],
      audioFilePath: json['audio_file_path'],
      originalAudioPath: json['original_audio_path'],
      translatedAudioPath: json['translated_audio_path'],
      sttConfidence: json['stt_confidence'],
      translationQuality: json['translation_quality'],
      ttsMethod: json['tts_method'],
      processingTimeMs: json['processing_time_ms'],
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : DateTime.now(),
      speakerName: json['speaker_name'],
      speakerAvatarUrl: json['speaker_avatar_url'],
    );
  }

  /// Create CallTranscript from timeline format (simplified)
  factory CallTranscript.fromTimelineJson(Map<String, dynamic> json) {
    return CallTranscript(
      id: json['id'] ?? '',
      callId: json['call_id'] ?? '',
      speakerUserId: json['speaker_id'],
      originalLanguage: json['language'] ?? 'he',
      originalText: json['text'] ?? '',
      translatedText: json['translated'],
      timestampMs: json['timestamp_ms'],
      createdAt: DateTime.now(),
    );
  }

  /// Convert CallTranscript to JSON
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'call_id': callId,
      'speaker_user_id': speakerUserId,
      'original_language': originalLanguage,
      'original_text': originalText,
      'translated_text': translatedText,
      'target_language': targetLanguage,
      'timestamp_ms': timestampMs,
      'audio_file_path': audioFilePath,
      'original_audio_path': originalAudioPath,
      'translated_audio_path': translatedAudioPath,
      'stt_confidence': sttConfidence,
      'translation_quality': translationQuality,
      'tts_method': ttsMethod,
      'processing_time_ms': processingTimeMs,
      'created_at': createdAt.toIso8601String(),
      'speaker_name': speakerName,
      'speaker_avatar_url': speakerAvatarUrl,
    };
  }

  /// Convert to timeline format
  Map<String, dynamic> toTimelineJson() {
    return {
      'timestamp_ms': timestampMs,
      'speaker_id': speakerUserId,
      'text': originalText,
      'translated': translatedText,
      'language': originalLanguage,
    };
  }

  /// Create a copy with updated fields
  CallTranscript copyWith({
    String? id,
    String? callId,
    String? speakerUserId,
    String? originalLanguage,
    String? originalText,
    String? translatedText,
    String? targetLanguage,
    int? timestampMs,
    String? audioFilePath,
    String? originalAudioPath,
    String? translatedAudioPath,
    int? sttConfidence,
    int? translationQuality,
    String? ttsMethod,
    int? processingTimeMs,
    DateTime? createdAt,
    String? speakerName,
    String? speakerAvatarUrl,
  }) {
    return CallTranscript(
      id: id ?? this.id,
      callId: callId ?? this.callId,
      speakerUserId: speakerUserId ?? this.speakerUserId,
      originalLanguage: originalLanguage ?? this.originalLanguage,
      originalText: originalText ?? this.originalText,
      translatedText: translatedText ?? this.translatedText,
      targetLanguage: targetLanguage ?? this.targetLanguage,
      timestampMs: timestampMs ?? this.timestampMs,
      audioFilePath: audioFilePath ?? this.audioFilePath,
      originalAudioPath: originalAudioPath ?? this.originalAudioPath,
      translatedAudioPath: translatedAudioPath ?? this.translatedAudioPath,
      sttConfidence: sttConfidence ?? this.sttConfidence,
      translationQuality: translationQuality ?? this.translationQuality,
      ttsMethod: ttsMethod ?? this.ttsMethod,
      processingTimeMs: processingTimeMs ?? this.processingTimeMs,
      createdAt: createdAt ?? this.createdAt,
      speakerName: speakerName ?? this.speakerName,
      speakerAvatarUrl: speakerAvatarUrl ?? this.speakerAvatarUrl,
    );
  }

  /// Check if translation was performed
  bool get hasTranslation => translatedText != null && translatedText!.isNotEmpty;

  /// Get display text (translated if available, otherwise original)
  String get displayText => translatedText ?? originalText;

  /// Get language display name
  String get languageDisplay {
    switch (originalLanguage) {
      case 'he':
        return 'Hebrew';
      case 'en':
        return 'English';
      case 'ru':
        return 'Russian';
      default:
        return originalLanguage;
    }
  }

  /// Get timestamp as formatted string
  String get timestampDisplay {
    if (timestampMs == null) return '--:--';
    final totalSeconds = timestampMs! ~/ 1000;
    final minutes = (totalSeconds ~/ 60).toString().padLeft(2, '0');
    final seconds = (totalSeconds % 60).toString().padLeft(2, '0');
    return '$minutes:$seconds';
  }

  /// Get TTS method display
  String get ttsMethodDisplay {
    switch (ttsMethod?.toLowerCase()) {
      case 'xtts':
      case 'xtts_voice_clone':
        return 'Voice Clone';
      case 'google_tts':
      case 'google_tts_fallback':
        return 'Standard TTS';
      case 'passthrough':
        return 'Direct Audio';
      default:
        return ttsMethod ?? 'Unknown';
    }
  }

  /// Get confidence display
  String get confidenceDisplay {
    if (sttConfidence == null) return 'N/A';
    return '$sttConfidence%';
  }
}

File: backend/app/models/call_transcript.py
"""
CallTranscript Model - Call History & Transcription

Stores complete word-by-word record of all calls for history.
"""
from sqlalchemy import Column, String, DateTime, Integer, Text, ForeignKey
from datetime import datetime, UTC
import uuid

from .database import Base


class CallTranscript(Base):
    """Transcript entry for a call"""
    __tablename__ = "call_transcripts"
    
    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    
    # References
    call_id = Column(String(36), ForeignKey('calls.id', ondelete='CASCADE'), nullable=False, index=True)
    speaker_user_id = Column(String(36), ForeignKey('users.id', ondelete='SET NULL'), nullable=True, index=True)
    
    # Content
    original_language = Column(String(10), nullable=False)
    original_text = Column(Text, nullable=False)
    translated_text = Column(Text, nullable=True)
    
    # Timing (milliseconds from call start)
    timestamp_ms = Column(Integer, nullable=True)
    
    # Timestamps
    created_at = Column(DateTime, default=lambda: datetime.utcnow(), nullable=False)
    
    @classmethod
    def create_transcript(
        cls,
        call_id: str,
        speaker_user_id: str,
        original_language: str,
        original_text: str,
        timestamp_ms: int,
        translated_text: str = None,
        **kwargs  # Ignore extra params (target_language, tts_method, etc.)
    ):
        """
        Factory method to create a transcript entry.
        
        Args:
            call_id: ID of the call
            speaker_user_id: ID of the speaker
            original_language: Language of original speech
            original_text: Transcribed text
            timestamp_ms: Timestamp in milliseconds from call start
            translated_text: Translated text (optional)
            **kwargs: Additional parameters (ignored for now)
            
        Returns:
            CallTranscript instance
        """
        return cls(
            call_id=call_id,
            speaker_user_id=speaker_user_id,
            original_language=original_language,
            original_text=original_text,
            translated_text=translated_text,
            timestamp_ms=timestamp_ms
        )

    
    def to_dict(self):
        return {
            "id": self.id,
            "call_id": self.call_id,
            "speaker_user_id": self.speaker_user_id,
            "original_language": self.original_language,
            "original_text": self.original_text,
            "translated_text": self.translated_text,
            "timestamp_ms": self.timestamp_ms,
            "created_at": self.created_at.isoformat() if self.created_at else None,
        }
